airbyte-config/init/src/main/java/io/airbyte/config/init/RemoteConnectorCatalogPersistence.java

```diff
@@ -0,0 +1,225 @@
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.commons.util.MoreIterators;
+import io.airbyte.config.AirbyteConfig;
+import io.airbyte.config.AirbyteConfigValidator;
+import io.airbyte.config.ConfigSchema;
+import io.airbyte.config.ConfigWithMetadata;
+import io.airbyte.config.persistence.ConfigNotFoundException;
+import io.airbyte.config.persistence.ConfigPersistence;
+import io.airbyte.validation.json.JsonValidationException;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This config persistence pulls the connector configurations from a remotely hosted catalog. It is
+ * read-only.
+ */
+final public class RemoteConnectorCatalogPersistence implements ConfigPersistence {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RemoteConnectorCatalogPersistence.class);
+  private static final String PERSISTENCE_READ_ONLY_ERROR_MSG = "The connector catalog is read only.";
+  private static final HttpClient httpClient = HttpClient.newHttpClient();
+
+  private static final Map<AirbyteConfig, String> CONFIG_SCHEMA_CONNECTOR_CATALOG_MAP = Map.of(
+      ConfigSchema.STANDARD_SOURCE_DEFINITION, "sources",
+      ConfigSchema.STANDARD_DESTINATION_DEFINITION, "destinations");
+
+  private final URI remoteConnectorCatalogUrl;
+  private static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(30);
+  private final Duration timeout;
+  // A mapping from config type to config UUID to config.
+  private ImmutableMap<AirbyteConfig, Map<String, JsonNode>> connectorCatalog;
+
+  public RemoteConnectorCatalogPersistence(final String remoteConnectorCatalogUrl) throws InterruptedException {
+    this.remoteConnectorCatalogUrl = URI.create(remoteConnectorCatalogUrl);
+    this.timeout = DEFAULT_TIMEOUT;
+    // TODO remove this call once dependency injection framework manages object creation
+    initialize();
+  }
+
+  public RemoteConnectorCatalogPersistence(final String remoteConnectorCatalogUrl, final Duration timeout) throws InterruptedException {
+    this.remoteConnectorCatalogUrl = URI.create(remoteConnectorCatalogUrl);
+    this.timeout = timeout;
+    // TODO remove this call once dependency injection framework manages object creation
+    initialize();
+  }
+
+  // TODO will be called automatically by the dependency injection framework on object creation
+  public void initialize() throws InterruptedException {
+    try {
+      final JsonNode rawConnectorCatalog = getRemoteConnectorCatalog(this.remoteConnectorCatalogUrl, this.timeout);
+      this.connectorCatalog = parseConnectorCatalog(rawConnectorCatalog);
```

Comment:
This feels like something we'd like to refresh. I'd expect this `RemoteCatalogConfigPersistence` should be a long lived object that would be keeping a cache of the remote catalog to avoid fetching this too frequently.

Not sure if this is the one that fits our usage but the idea would be to rely on something like a [CacheLoader](https://guava.dev/releases/21.0/api/docs/com/google/common/cache/Cache.html).